popovertarget on button type=button in form doesn't work

lukewarlow opened the pull request:
#### 9a8ecbaa8fc59d6f4336fffe04c31986cdc5e92e
<pre>
popovertarget on button type=button in form doesn&apos;t work
<a href="https://bugs.webkit.org/show_bug.cgi?id=275899">https://bugs.webkit.org/show_bug.cgi?id=275899</a>

Reviewed by Darin Adler.

This patch fixes an issue where buttons with popovertarget never work when inside a form.

* LayoutTests/imported/w3c/web-platform-tests/html/semantics/popovers/button-type-popovertarget-expected.txt: Added.
* LayoutTests/imported/w3c/web-platform-tests/html/semantics/popovers/button-type-popovertarget.html: Added.
* LayoutTests/imported/w3c/web-platform-tests/html/semantics/popovers/button-type-reset-popovertarget.tentative-expected.txt: Removed.
* LayoutTests/imported/w3c/web-platform-tests/html/semantics/popovers/button-type-reset-popovertarget.tentative.html: Removed.
* Source/WebCore/html/HTMLButtonElement.cpp:
(WebCore::HTMLButtonElement::defaultEventHandler):

Canonical link: <a href="https://commits.webkit.org/282801@main">https://commits.webkit.org/282801@main</a>
</pre>
<!--EWS-Status-Bubble-Start-->
https://github.com/WebKit/WebKit/commit/053923c5f4e93e192f9cc13821bf27fe0871fa17

| Misc | iOS, visionOS, tvOS & watchOS  | macOS  | Linux |  Windows |
| ----- | ---------------------- | ------- |  ----- |  --------- |
| [✅ 🧪 style](https://ews-build.webkit.org/#/builders/38/builds/64310 "Passed style check") | [✅ 🛠 ios](https://ews-build.webkit.org/#/builders/48/builds/43667 "Built successfully") | [✅ 🛠 mac](https://ews-build.webkit.org/#/builders/55/builds/16907 "Built successfully") | [✅ 🛠 wpe](https://ews-build.webkit.org/#/builders/5/builds/68332 "Built successfully") | [✅ 🛠 wincairo](https://ews-build.webkit.org/#/builders/59/builds/14918 "Built successfully") 
| [✅ 🧪 bindings](https://ews-build.webkit.org/#/builders/9/builds/66430 "Passed tests") | [✅ 🛠 ios-sim](https://ews-build.webkit.org/#/builders/49/builds/51365 "Built successfully") | [✅ 🛠 mac-AS-debug](https://ews-build.webkit.org/#/builders/61/builds/15198 "Built successfully") | [✅ 🧪 wpe-wk2](https://ews-build.webkit.org/#/builders/34/builds/51755 "Passed tests") | [✅ 🧪 wincairo-tests](https://ews-build.webkit.org/#/builders/60/builds/10294 "Passed tests") 
| [✅ 🧪 webkitperl](https://ews-build.webkit.org/#/builders/11/builds/67379 "Passed tests") | [✅ 🧪 ios-wk2](https://ews-build.webkit.org/#/builders/47/builds/40371 "Passed tests") | [✅ 🧪 api-mac](https://ews-build.webkit.org/#/builders/18/builds/55648 "Passed tests") | [✅ 🧪 api-wpe](https://ews-build.webkit.org/#/builders/41/builds/32374 "Passed tests") | 
| | [✅ 🧪 ios-wk2-wpt](https://ews-build.webkit.org/#/builders/42/builds/37041 "Passed tests") | [✅ 🧪 mac-wk1](https://ews-build.webkit.org/#/builders/64/builds/13027 "Passed tests") | [✅ 🛠 wpe-cairo](https://ews-build.webkit.org/#/builders/65/builds/13792 "Built successfully") | 
| | [✅ 🧪 api-ios](https://ews-build.webkit.org/#/builders/13/builds/59007 "Passed tests") | [✅ 🧪 mac-wk2](https://ews-build.webkit.org/#/builders/63/builds/13356 "Passed tests") | [✅ 🛠 gtk](https://ews-build.webkit.org/#/builders/2/builds/70031 "Built successfully") | 
| | [✅ 🛠 vision](https://ews-build.webkit.org/#/builders/87/builds/8257 "Built successfully") | [✅ 🧪 mac-AS-debug-wk2](https://ews-build.webkit.org/#/builders/62/builds/12873 "Passed tests") | [✅ 🧪 gtk-wk2](https://ews-build.webkit.org/#/builders/1/builds/59086 "Passed tests") | 
| | [✅ 🛠 vision-sim](https://ews-build.webkit.org/#/builders/86/builds/8290 "Built successfully") | [✅ 🧪 mac-wk2-stress](https://ews-build.webkit.org/#/builders/8/builds/55739 "Passed tests") | [✅ 🧪 api-gtk](https://ews-build.webkit.org/#/builders/21/builds/59241 "Passed tests") | 
| [✅ 🛠 🧪 merge](https://ews-build.webkit.org/#/builders/19/builds/14196 "Built successfully and passed tests") | [✅ 🧪 vision-wk2](https://ews-build.webkit.org/#/builders/88/builds/6828 "Passed tests") | [✅ 🧪 mac-intel-wk2](https://ews-build.webkit.org/#/builders/119/builds/514 "Passed tests") | | 
| | [✅ 🛠 tv](https://ews-build.webkit.org/#/builders/44/builds/39487 "Built successfully") | | | 
| | [✅ 🛠 tv-sim](https://ews-build.webkit.org/#/builders/46/builds/40566 "Built successfully") | | | 
| | [✅ 🛠 watch](https://ews-build.webkit.org/#/builders/43/builds/41749 "Built successfully") | | | 
| | [✅ 🛠 watch-sim](https://ews-build.webkit.org/#/builders/45/builds/40309 "Built successfully") | | | 
<!--EWS-Status-Bubble-End-->

lukewarlow commented on Source/WebCore/html/HTMLButtonElement.cpp:
```diff
@@ -155,10 +155,15 @@ void HTMLButtonElement::defaultEventHandler(Event& event)
-            handlePopoverTargetAction();
+        }
+
+        if (!protectedForm || m_type == BUTTON) {
```
Marking as draft because this maybe should allow reset buttons too.

darinadler replied:
Does the WPT cover this and answer that question? If not, can we add a test rather than just deciding based on our thinking?

lukewarlow replied:
Apologies I should have resolved this comment. The spec (and Firefox) both support reset buttons so I've made them work in WebKit and updated the wpt to match.

annevk commented on Source/WebCore/html/HTMLButtonElement.cpp:
```diff
@@ -155,10 +155,15 @@ void HTMLButtonElement::defaultEventHandler(Event& event)
-            handlePopoverTargetAction();
+        }
+
+        if (!(protectedForm && m_type == SUBMIT)) {
```
Why not use else if here?

lukewarlow replied:
Because that's eactly the bug we currently have. If you use else if you won't get it working inside forms?

darinadler commented on Source/WebCore/html/HTMLButtonElement.cpp:
```diff
@@ -155,10 +155,15 @@ void HTMLButtonElement::defaultEventHandler(Event& event)
+            if (invokeTargetElement())
+                handleInvokeAction();
+            else
+                handlePopoverTargetAction();
```
Do we need to call `setDefaultHandled` for any edges cases? Can we add tests that cover that explicitly?

lukewarlow replied:
Included as part of follow up bug

darinadler reviewed (approved):
This change is likely correct, but test coverage should be as fine grained as possible so we know there is consistency in the different cases, with and without a form, different types, with an without a target element, and even "if the side effect of the event dispatched as part of submit changes whether the element is still in the form or changes the type". I think we may need more tests to actually achieve interoperability, rather than just fixing everything needed to pass the tests currently in this pull request.

lukewarlow committed:
popovertarget on button type=button in form doesn't work
https://bugs.webkit.org/show_bug.cgi?id=275899

Reviewed by Darin Adler.

This patch fixes an issue where buttons with popovertarget never work when inside a form.

* LayoutTests/imported/w3c/web-platform-tests/html/semantics/popovers/button-type-popovertarget-expected.txt: Added.
* LayoutTests/imported/w3c/web-platform-tests/html/semantics/popovers/button-type-popovertarget.html: Added.
* LayoutTests/imported/w3c/web-platform-tests/html/semantics/popovers/button-type-reset-popovertarget.tentative-expected.txt: Removed.
* LayoutTests/imported/w3c/web-platform-tests/html/semantics/popovers/button-type-reset-popovertarget.tentative.html: Removed.
* Source/WebCore/html/HTMLButtonElement.cpp:
(WebCore::HTMLButtonElement::defaultEventHandler):

Canonical link: https://commits.webkit.org/282801@main
